File: src/nlp/intent_classifier.py
from dataclasses import dataclass
from typing import List, Dict, Literal
import re

IntentType = Literal[
    'describe_symptoms',
    'ask_recommendations',
    'emergency_help',
    'clarify_symptoms',
    'provide_history',
    'unknown'
]

@dataclass
class IntentClassification:
    """Classification result for user intent."""
    intent: IntentType
    confidence: float
    sub_intents: List[str]

class IntentClassifier:
    """Classifies user intent based on input text."""

    INTENT_PATTERNS: Dict[str, List[str]] = {
        'describe_symptoms': [
            'feel', 'having', 'experiencing', 'suffering',
            'pain', 'ache', 'hurts', 'symptoms'
        ],
        'ask_recommendations': [
            'what should', 'how can', 'help', 'advice',
            'recommend', 'suggestion', 'treatment'
        ],
        'emergency_help': [
            'emergency', 'urgent', 'immediately', 'severe',
            'worst', 'unbearable', 'help'
        ],
        'clarify_symptoms': [
            'mean', 'explain', 'clarify', 'specifically',
            'exactly', 'detail'
        ],
        'provide_history': [
            'happened', 'started', 'began', 'history',
            'before', 'previous', 'past'
        ]
    }

    def classify_intent(self, text: str) -> IntentClassification:
        """
        Classify the intent of user input text.
        
        Args:
            text: User input text
            
        Returns:
            IntentClassification object with intent and confidence
        """
        normalized_text = text.lower()
        max_confidence = 0.0
        primary_intent: IntentType = 'unknown'
        sub_intents: List[str] = []

        for intent, patterns in self.INTENT_PATTERNS.items():
            confidence = self._calculate_intent_confidence(normalized_text, patterns)
            
            if confidence > max_confidence:
                max_confidence = confidence
                primary_intent = intent  # type: ignore

            if confidence > 0.3:
                sub_intents.append(intent)

        return IntentClassification(
            intent=primary_intent,
            confidence=max_confidence,
            sub_intents=sub_intents
        )

    def _calculate_intent_confidence(self, text: str, patterns: List[str]) -> float:
        matches = 0
        total_weight = 0.0

        for pattern in patterns:
            regex = re.compile(rf'\b{pattern}\b', re.I)
            if match := regex.search(text):
                matches += 1
                
                # Weight based on position
                position = match.start()
                position_weight = 1 - (position / len(text))
                total_weight += position_weight

        pattern_count = len(patterns)
        return (matches / pattern_count) * 0.7 + (total_weight / pattern_count) * 0.3

File: src/lib/nlp/intentClassifier.ts
export type Intent = 
  | 'describe_symptoms'
  | 'ask_recommendations'
  | 'emergency_help'
  | 'clarify_symptoms'
  | 'provide_history'
  | 'unknown';

export interface IntentClassification {
  intent: Intent;
  confidence: number;
  subIntents: string[];
}

export class IntentClassifier {
  private static readonly INTENT_PATTERNS = {
    describe_symptoms: [
      'feel', 'having', 'experiencing', 'suffering',
      'pain', 'ache', 'hurts', 'symptoms'
    ],
    ask_recommendations: [
      'what should', 'how can', 'help', 'advice',
      'recommend', 'suggestion', 'treatment'
    ],
    emergency_help: [
      'emergency', 'urgent', 'immediately', 'severe',
      'worst', 'unbearable', 'help'
    ],
    clarify_symptoms: [
      'mean', 'explain', 'clarify', 'specifically',
      'exactly', 'detail'
    ],
    provide_history: [
      'happened', 'started', 'began', 'history',
      'before', 'previous', 'past'
    ]
  };

  classifyIntent(text: string): IntentClassification {
    const normalizedText = text.toLowerCase();
    let maxConfidence = 0;
    let primaryIntent: Intent = 'unknown';
    const subIntents: string[] = [];

    // Calculate confidence for each intent
    for (const [intent, patterns] of Object.entries(IntentClassifier.INTENT_PATTERNS)) {
      const confidence = this.calculateIntentConfidence(normalizedText, patterns);
      
      if (confidence > maxConfidence) {
        maxConfidence = confidence;
        primaryIntent = intent as Intent;
      }
      
      if (confidence > 0.3) {
        subIntents.push(intent);
      }
    }

    return {
      intent: primaryIntent,
      confidence: maxConfidence,
      subIntents: subIntents
    };
  }

  private calculateIntentConfidence(text: string, patterns: string[]): number {
    let matches = 0;
    let totalWeight = 0;

    patterns.forEach(pattern => {
      const regex = new RegExp(`\\b${pattern}\\b`, 'i');
      if (regex.test(text)) {
        matches++;
        
        // Add weight based on pattern position
        const position = text.search(regex);
        const positionWeight = 1 - (position / text.length);
        totalWeight += positionWeight;
      }
    });

    return (matches / patterns.length) * 0.7 + (totalWeight / patterns.length) * 0.3;
  }
}